wlutil/build.py

```diff
@@ -186,23 +186,78 @@ def buildWorkload(cfgName, cfgs, buildBin=True, buildImg=True):
     if ret != 0:
         raise RuntimeError("Error while building workload")
 
+def makeDrivers(boardDir, linuxSrc):
+    driverDirs = pathlib.Path(boardDir).glob("drivers/*")
+    makeCmd = "make -C " + str(linuxSrc) + " ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- "
+
+    # Prepare the linux source for building external drivers
+    run(["make", "ARCH=riscv", "CROSS_COMPILE=riscv64-unknown-linux-gnu-", "modules_prepare", jlevel], cwd=linuxSrc)
+
+    drivers = []
+    for driverDir in driverDirs:
+        run(makeCmd + "M=" + str(driverDir), cwd=driverDir, shell=True)
+        drivers.extend(list(driverDir.glob("*.ko")))
+
+    return drivers
+
+def setupBoardInitramfs(boardDir, linuxSrc):
+    log = logging.getLogger()
+
+    loadScript = initramfs_root / "loadDrivers.sh"
+
+    # clear out any leftover drivers from a previous build
+    for oldDriver in initramfs_root.glob("*.ko"):
+        oldDriver.unlink()
+    if loadScript.exists():
+        loadScript.unlink()
+
+    drivers = makeDrivers(boardDir, linuxSrc)
+
+    for driverPath in drivers:
+        shutil.copy(driverPath, initramfs_root)
+
+    with open(loadScript, 'w') as f:
+        for driverPath in drivers:
+            f.write("insmod /" + str(driverPath.name) + "\n")
+
+    if not (initramfs_root / "bin" / "busybox").exists():
+        raise RunTimeError("Initramfs root not initialized. Please run './marshal init'")
+
+    # This script adds the device nodes and packages up the initrd.
+    run(['fakeroot', '--', './makeInitramfs.sh'], cwd=wlutil_dir)
+
 # Now build linux/bbl
 def makeBin(config, initramfs=False):
     log = logging.getLogger()
 
     # We assume that if you're not building linux, then the image is pre-built (e.g. during host-init)
     if 'linux-config' in config:
         linuxCfg = os.path.join(config['linux-src'], '.config')
-        shutil.copy(config['linux-config'], linuxCfg)
+        defCfg = os.path.join(config['linux-src'], 'defconfig')
+        # Create a defconfig to use as reference
+        if not os.path.isfile(defCfg):
+            run(['make', 'ARCH=riscv', 'defconfig'], cwd=config['linux-src'])
+            shutil.copy(linuxCfg, defCfg)
+            with open(defCfg, 'a') as f:
+                f.write("CONFIG_BLK_DEV_INITRD=y\n")
+                f.write('CONFIG_INITRAMFS_SOURCE="' + str(initramfs_cpio) + '"\n')
+
+        # Create a config from the user fragments
+        kconfigEnv = os.environ.copy()
+        kconfigEnv['ARCH'] = 'riscv'
+        run([os.path.join(config['linux-src'], 'scripts/kconfig/merge_config.sh'),
```

Comment:
In the future, I'm going to change this so that users can provide an arbitrary list of kconfig fragments and the configs can stack on parent workloads. I'm not doing this yet to keep the functional changes to a minimum in this PR.